src/renci_ner/services/normalization/nodenorm.py

```diff
@@ -0,0 +1,144 @@
+#
+# The Translator Node Normalizer as a Transformer.
+# Source code: https://github.com/TranslatorSRI/NodeNormalization
+# Hosted at: https://nodenormalization-sri.renci.org/
+#
+import logging
+
+import requests
+
+from renci_ner.core import (
+    AnnotatedText,
+    AnnotationProvenance,
+    NormalizedAnnotation,
+    Transformer,
+)
+
+# Configuration.
+RENCI_NODENORM_URL = "https://nodenormalization-sri.renci.org"
+
+
+class NodeNorm(Transformer):
+    """
+    The Translator Node Normalizer as a Transformer.
+    """
+
+    @property
+    def provenance(self) -> AnnotationProvenance:
+        """Return an AnnotationProvenance describing annotations produced by this service."""
+        return AnnotationProvenance(
+            name="NodeNorm", url=RENCI_NODENORM_URL, version=self.openapi_version
+        )
+
+    def __init__(self, url=RENCI_NODENORM_URL, requests_session=requests.Session()):
+        """
+        Set up a BioMegatron service.
+
+        :param url: The URL of the BioMegatron service.
+        :param requests_session: A Requests session object to use instead of the default one.
+        """
+        self.url = url
+        self.get_normalized_nodes_url = url + "/get_normalized_nodes"
+        self.requests_session = requests_session
+
+        response = requests.get(self.url + "/openapi.json")
+        response.raise_for_status()
+        openapi_data = response.json()
+        self.openapi_version = openapi_data.get("info", {"version": "NA"}).get(
+            "version", "NA"
+        )
+
+    def supported_properties(self):
+        """Some configurable parameters."""
+        return {
+            "geneprotein_conflation": "(true/false, default: true) Whether to conflate gene and protein identifiers.",
+            "drugchemical_conflation": "(true/false, default: false) Whether to conflate drug and chemical identifiers.",
+            "description": "(true/false, default: false) Whether to include descriptions in the response.",
+        }
+
+    def transform(self, annotated_text: AnnotatedText, props=None) -> AnnotatedText:
+        """
+        Transform an AnnotatedText object using NodeNorm. For every annotation, we pass the IDs to NodeNorm, and if
+        it changes the identifier, we would return a NormalizedAnnotation. Otherwise, we return the original
+        annotation.
+
+        :param annotated_text: The annotated text to transform.
+        :param props: Properties to pass to NodeNorm (see supported_properties).
+        :return: The AnnotatedText with normalized annotations where possible.
+        """
+        if props is None:
+            props = {}
+        session = self.requests_session
+
+        ids = list(map(lambda a: a.id, annotated_text.annotations))
```

Comment:
I assume ```ids``` could include duplicate identifiers. Would it be beneficial to deduplicate before sending it to the nodenorm for normalization?

Reply:
Fixed in 365e4ed.